base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java

```diff
@@ -0,0 +1,650 @@
+        }
+
+        String dn = ATTR_AUTHORIZATION_ID + "=" + authorization.getID()
+                        + "," + RDN_AUTHORIZATION + "," + basedn;
+        LDAPEntry entry = new LDAPEntry(dn, attrSet);
+        ldapAdd(entry);
+    }
+
+    public void addChallenge(String accountID, ACMEChallenge challenge)
+            throws Exception {
+        String type = challenge.getType();
+        String objclass = null;
+        if (type.equals("dns-01")) {
+            objclass = OBJ_CHALLENGE_DNS01;
+        } else if (type.equals("http-01")) {
+            objclass = OBJ_CHALLENGE_HTTP01;
+        } else {
+            throw new RuntimeException("unrecognised challenge type");
+        }
+
+        String[] classes = {OBJ_CHALLENGE, objclass};
+        LDAPAttribute[] attrs = {
+            new LDAPAttribute(ATTR_OBJECTCLASS, classes),
+            new LDAPAttribute(ATTR_CHALLENGE_ID, challenge.getID()),
+            new LDAPAttribute(ATTR_AUTHORIZATION_ID, challenge.getAuthzID()),
+            new LDAPAttribute(ATTR_ACCOUNT_ID, accountID),
+            new LDAPAttribute(ATTR_STATUS, challenge.getStatus()),
+            new LDAPAttribute(ATTR_TOKEN, challenge.getToken())
+        };
+        LDAPAttributeSet attrSet = new LDAPAttributeSet(attrs);
+
+        String dn = ATTR_CHALLENGE_ID + "=" + challenge.getID()
+                        + "," + RDN_CHALLENGE + "," + basedn;
+        LDAPEntry entry = new LDAPEntry(dn, attrSet);
+        ldapAdd(entry);
+    }
+
+    public void updateAuthorization(ACMEAuthorization authorization) throws Exception {
+        String dn = ATTR_AUTHORIZATION_ID + "=" + authorization.getID()
+                        + "," + RDN_AUTHORIZATION + "," + basedn;
+        LDAPModificationSet mods = new LDAPModificationSet();
+        mods.add(
+            LDAPModification.REPLACE,
+            new LDAPAttribute(ATTR_STATUS, authorization.getStatus())
+        );
+        ldapModify(dn, mods);
+
+        // delete and re-add challenge objects
+        for (ACMEChallenge challenge : authorization.getChallenges()) {
+            dn = ATTR_CHALLENGE_ID + "=" + challenge.getID()
+                    + "," + RDN_CHALLENGE + "," + basedn;
+            try {
+                ldapDelete(dn, OnNoSuchObject.Ignore);
+            } catch (LDAPException e) {
+                if (e.getLDAPResultCode() == LDAPException.NO_SUCH_OBJECT) {
+                    // ignore
+                } else {
+                    throw new Exception("LDAPDatabase.ldapDelete: failed", e);
+                }
+            }
+            addChallenge(authorization.getAccountID(), challenge);
+        }
+    }
+
+
+
+    /* LOW LEVEL LDAP METHODS */
+
+    void ldapAdd(LDAPEntry entry) throws Exception {
+        LDAPConnection conn = connFactory.getConn();
+        try {
+            conn.add(entry);
+        } catch (LDAPException e) {
+            throw new Exception("LDAPDatabase.ldapAdd: failed to add entry", e);
```

Comment:
Could we chain the exception message, e.g.:
```
throw new Exception("LDAPDatabase.ldapAdd: failed to add entry: " + e, e);
```
Same thing with the rest of the code below.